armi/reactor/converters/axialExpansionChanger/axialExpansionChanger.py

```diff
@@ -448,13 +448,48 @@ def axiallyExpandAssembly(self):
                     c.height = c.ztop - c.zbottom
 
             self._checkBlockHeight(b)
+            self._recomputeBlockMassParams(b)
+
             # redo mesh -- functionality based on assembly.calculateZCoords()
             mesh.append(b.p.ztop)
             b.spatialLocator = self.linked.a.spatialGrid[0, 0, ib]
 
+        # after expansion, recalculate all the burnups
+        for b in self.linked.a:
+            self._recalculateBurnup(b)
+
         bounds = list(self.linked.a.spatialGrid._bounds)
         bounds[2] = array(mesh)
         self.linked.a.spatialGrid._bounds = tuple(bounds)
+    
+    def _recalculateBurnup(self, b):
+        """After moving nuclides around, recalculate burnup."""
+        for c in b.iterComponents(Flags.FUEL):
+            c.p.percentBu = self._calcBurnup(c.getHMMoles(), c.p.molesHmBOL) 
+        b.p.percentBu = self._calcBurnup(b.getHMMoles(), b.p.molesHmBOL)
+
+    def _calcBurnup(self, currentHM, initialHM):
```

Comment:
@drewj-tp and I were planning on taking this internal with the extensibility improvements being done in #2289

Reply:
See https://github.com/terrapower/armi/pull/2289/commits/07fc6c1a0cb2287777466786f9c56d29730441e3. I'll also reach out offline

Reply:
Yeah not quite ready. We're chipping away at the block though!

Reply:
The tests pass!